### ISSUE TITLE
[question] private versions - what would be conan's recommendation of handling this?

### ISSUE THREAD

**Issue by fulara** [conan-io/conan#6569]
At my work environment unfortunatelly at the moment we cant test many things locally. 
We are now slowly migrating towards conan, however a question came up - how to deal with private versions.

Whats a private version?
Private version is a version of particular package that you dont want to upload to a remote artifactory, but you need to transfer it across hosts.
In our working environment currently we have nfs storage thats accessible on both build server and test server.

Historically our package repository was just filesystem, so in order to use private version, people just created symlinks into the nfs, and built packages into that same nfs, hence everytime when you built package on build server it was immediatelly available on test server.

On test server right now we have a conan storage thats pointing to our 'remote' repository, 
So if people have builds build by CI which pushes onto remote they can use standard conan procedure to install it. 

However question arose about private versions, basically when you are incrementally building your application on build server, and dont want to clutter remote with it, 
What would be the conan recommended way of getting this package into test servers' local cache without polluting remote?

At the moment we have two proposals:
1) rsync packages from build server cache onto test server local cache. 
First export packages onto build server local cache with help of `conan export-pkg`.
Then rsync them so something like:  ( writing from memory so dont worry about path errors. ) 
`rsync my_build_server:/.conan/data/test-project/1.0/test  /.conan/data/test-project/1.0/`
This seems to work and is probably the way we will be moving forward, 
However it seems slightly wrong as we are toying with conan data storage  by hand.
2) host an alternative remote, probably one per user and just add this remote on test-server.
So whoever would need to deploy their private test version, could just create their own remote, and fetch packages through that, probably we would go with using conan_server as we need a minimalistic aproach.
This aproach seems more conan friendly, however would require us to clutter our test server remotes with many additional endpoints.

Which  would you recommend? 
Or maybe you would have an alternative way of doing this?

**Comment by jgsogo** [comment]
Hi, @fulara 

And what about creating a _private_ repository in the Artifactory you already have instead of creating (and running) a `conan_server` per user? You can use the Artifactory API to create the repo and set the permissions for the required user, upload those private versions there and only the user will be able to access them (from any other node/server/machine in your organization). It will scale pretty well and doesn't require running more processes.

**Comment by fulara** [comment]
Hello @jgsogo 

I think my request together with: https://github.com/conan-io/conan/issues/6128 show that maybe conan should provide some way of migrating packages from one cache to another, (preferably across hosts. ) However since currently is basically as simple as just rsyncing from one to another maybe thats overkill.

Yes, repos in artifactory is something we have considered, I failed to mention that in my original post.

I think we will go with two options right now.
the '1)' - where we will let users rsync packages themselves, 
and a variant of yours, but will just create a single repository for non-prod stuff. 

I've tested that conan is happy accepting packages that are rsync'ed into its .conan - and I am happy that is the case - I was afraid that it would have baked into some db somewhere what packages it accepts.

I am slightly worried about multiple repos though, historically we had multiple repos but it was too hard to manage them, although that was a slightly different case though.
We had repos for 'internal packages' and 'ig packages', that coupled with usage of `upload '*'` caused loads of issue because of the way conan did search for packages at the time. 
(If conan found at least one package under name/version... in first repository it didnt bother to look into next repositories ) 

Anyway, enough of my bubbling, thanks for your response. I'll close this one as I dont think any better aproach is possible at the moment.
